src/sensors.c

```diff
@@ -11,63 +11,198 @@ extern "C" {
 #include "mavlink_log.h"
 #include "param.h"
 #include "sensors.h"
+#include "estimator.h"
+#include "mode.h"
 
+#include "turbotrig/turbovec.h"
+
+//==================================================================
 // global variable definitions
-bool _imu_ready;
+
+// IMU
 vector_t _accel;
 vector_t _gyro;
 float _imu_temperature;
-uint32_t _imu_time;
+uint64_t _imu_time;
+bool new_imu_data = false;
+
+// Airspeed
+bool _diff_pressure_present = false;
+float _pitot_velocity, _pitot_diff_pressure, _pitot_temp;
 
-bool _diff_pressure_present;
-int16_t _diff_pressure;
-int16_t _temperature;
+// Barometer
+bool _baro_present = false;
+float _baro_altitude;
+float _baro_pressure;
+float _baro_temperature;
 
-bool _baro_present;
-int16_t _baro_pressure;
-int16_t _baro_temperature;
+// Sonar
+bool _sonar_present = false;
+float _sonar_range;
 
-bool _sonar_present;
-int16_t _sonar_range;
-uint32_t _sonar_time;
+// Magnetometer
+bool _mag_present = false;
+vector_t _mag;
 
+
+//==================================================================
 // local variable definitions
-static uint32_t diff_press_next_us;
-static uint32_t baro_next_us;
-static uint32_t sonar_next_us;
 
+// IMU stuff
+int16_t accel_raw[3];
+int16_t gyro_raw[3];
+int16_t temp_raw;
+static volatile uint8_t accel_status, gyro_status, temp_status;
 static float accel_scale;
 static float gyro_scale;
+static bool calibrating_acc_flag;
+static bool calibrating_gyro_flag;
+static void calibrate_accel(void);
+static void calibrate_gyro(void);
+static void correct_imu(void);
+static void imu_ISR(void);
+static bool update_imu(void);
+
+
+//==================================================================
+// function definitions
+void init_sensors(void)
+{
+  while(millis() < 50);
+  i2cWrite(0,0,0);
+  _baro_present = ms5611_init();
+  _mag_present = hmc5883lInit(get_param_int(PARAM_BOARD_REVISION));
+  _sonar_present = mb1242_init();
+  _diff_pressure_present = ms4525_init();
+
+  // IMU
+  uint16_t acc1G;
+  mpu6050_init(true, &acc1G, &gyro_scale, get_param_int(PARAM_BOARD_REVISION));
+  mpu6050_register_interrupt_cb(&imu_ISR);
+  accel_scale = 9.80665f/acc1G * get_param_float(PARAM_ACCEL_SCALE);
+}
 
-static bool calib_gyro;
-static bool calib_acc;
 
+bool update_sensors()
+{
+  // Look for disabled sensors while disarmed (poll every 10 seconds)
+  // These sensors need power to respond, so they might not have been
+  // detected on startup, but will be detected whenever power is applied
+  // to the 5V rail.
```

Comment:
Cool. Do you think we should up that rate a bit? This shouldn't take all that much extra time